### PR TITLE
TOOL-527 Configuration can handle range values.

### PR DESCRIPTION
**Pull request by richard-bogdan-bitrise**

### CODE REVIEW

**Comment by richard-bogdan-bitrise** [comment]
Updated code accordingly. Please check.

Note: for exclusive inclusive please see https://en.wikipedia.org/wiki/Bracket_(mathematics) Interval sections

File: stepconf/stepconf.go [hunk 1]
```diff
@@ -135,11 +137,15 @@ func setField(field reflect.Value, value, constraint string) error {
 			return err
 		}
 	// TODO: use FindStringSubmatch to distinguish no match and match for empty string.
-	case regexp.MustCompile(`^opt\[.*\]$`).FindString(constraint):
+	case regexp.MustCompile(`^opt\[.*]$`).FindString(constraint):
```
[the remark this answers is not in the document]
Comment (richard-bogdan-bitrise) [reply stepconf/stepconf.go]: Yes, the reason was that it was a redundant char, re-added the escape char to maintain readability

File: stepconf/stepconf.go [hunk 2]
```diff
@@ -171,6 +177,229 @@ func setField(field reflect.Value, value, constraint string) error {
 	return nil
 }
 
+//ValidateRangeFields validates if the given range is proper.
+func ValidateRangeFields(valueStr, constraint string) error {
+	var constraintMin string
```
[the remark this answers is not in the document]
Comment (richard-bogdan-bitrise) [reply stepconf/stepconf.go]: Merged.

File: stepconf/stepconf.go [hunk 3]
```diff
@@ -171,6 +177,229 @@ func setField(field reflect.Value, value, constraint string) error {
...
+	var constraintMax string
+	var err error
+
+	if constraintMin, constraintMax, err = GetRangeValues(constraint); err != nil {
```
[the remark this answers is not in the document]
Comment (richard-bogdan-bitrise) [reply stepconf/stepconf.go]: Removed declarations.

File: stepconf/stepconf.go [hunk 4]
```diff
@@ -171,6 +177,229 @@ func setField(field reflect.Value, value, constraint string) error {
...
+		return err
+	}
+	var min interface{}
+	if min, err = parseValueStr(constraintMin); err != nil {
```
[the remark this answers is not in the document]
Comment (richard-bogdan-bitrise) [reply stepconf/stepconf.go]: Updated ode.

File: stepconf/stepconf.go [hunk 5]
```diff
@@ -140,6 +142,10 @@ func setField(field reflect.Value, value, constraint string) error {
 			// TODO: print only the value options, not the whole string.
 			return fmt.Errorf("value is not in value options (%s)", constraint)
 		}
+	case regexp.MustCompile(rangeRegex).FindString(constraint):
+		if err := ValidateRangeFields(value, constraint); err != nil {
+			return err
+		}
 	default:
 		return fmt.Errorf("invalid constraint (%s)", constraint)
 	}
```
Comment (trapacska) [inline stepconf/stepconf.go]: from line 157, please include float parsing

File: stepconf/stepconf.go [hunk 6]
```diff
@@ -171,6 +177,227 @@ func setField(field reflect.Value, value, constraint string) error {
...
+	return nil
+}
+
+// GetRangeValues reads up the given range constraint and returns the values, or an error if the constraint is malformed or could not be parsed.
+func GetRangeValues(value string) (string, string, error) {
+	groups := regexp.MustCompile(rangeRegex).FindAllStringSubmatch(value, len(value))
+	if len(groups) < 1 {
+		return "", "", fmt.Errorf("value in value options is malformed (%s)", value)
+	}
+	minStr := getMinValueStr(groups)
```
Comment (trapacska) [inline stepconf/stepconf.go]: this and the next line is can be just groups[0] and groups[1]?

File: stepconf/stepconf.go [hunk 7]
```diff
@@ -171,6 +177,227 @@ func setField(field reflect.Value, value, constraint string) error {
...
+		}
+		return getFloatValue(parsedValue)
+	default:
+		return 0, fmt.Errorf("not supported type %T", value)
+	}
+}
+
+func validateRangeMinFieldValue(min float64, value float64) error {
+	if min > value {
+		return fmt.Errorf("value %f is out of range, lesser than minimum %f", value, min)
```
Comment (trapacska) [inline stepconf/stepconf.go]: lesser -> less

File: stepconf/stepconf.go (hunk 6 above)
Comment (trapacska) [reply stepconf/stepconf.go]: not sure if we need FindAllStringSubmatch here, also, if we have named groups then we should select the group by name

File: stepconf/stepconf.go (hunk 7 above)
Comment (richard-bogdan-bitrise) [reply stepconf/stepconf.go]: Right :)

File: stepconf/stepconf.go (hunk 5 above)
Comment (richard-bogdan-bitrise) [reply stepconf/stepconf.go]: Added case float.

File: stepconf/stepconf.go (hunk 6 above)
Comment (richard-bogdan-bitrise) [reply stepconf/stepconf.go]: Updated code to use named group matching.

File: stepconf/stepconf.go [hunk 8]
```diff
@@ -171,6 +187,260 @@ func setField(field reflect.Value, value, constraint string) error {
...
+	}
+	if err := validateRangeFieldTypes(min, max, value); err != nil {
+		return err
+	}
+
+	return nil
+}
+
+func isMinInclusive(bracket string) (bool, error) {
+	if bracket == "[" {
```
Comment (trapacska) [inline stepconf/stepconf.go]: please use a switch with a default element so you no need to have a return at the end

File: stepconf/stepconf.go [hunk 9]
```diff
@@ -171,6 +187,260 @@ func setField(field reflect.Value, value, constraint string) error {
...
+		return true, nil
+	} else if bracket == "]" {
+		return false, nil
+	}
+	return false, fmt.Errorf("invalid string found for bracket: %s", bracket)
+}
+
+func isMaxInclusive(bracket string) (bool, error) {
```
Comment (trapacska) [inline stepconf/stepconf.go]: same as above

File: stepconf/stepconf.go [hunk 10]
```diff
@@ -171,6 +187,260 @@ func setField(field reflect.Value, value, constraint string) error {
...
+		}
+		return getFloatValue(parsedValue)
+	default:
+		return 0, fmt.Errorf("not supported type %T", value)
+	}
+}
+
+func validateRangeMinFieldValue(min float64, value float64, inclusive bool) error {
+	if inclusive {
+		if min > value {
```
Comment (trapacska) [inline stepconf/stepconf.go]: please merge the inner ifs into the main ones

File: stepconf/stepconf.go [hunk 11]
```diff
@@ -171,6 +187,260 @@ func setField(field reflect.Value, value, constraint string) error {
...
+		if min >= value {
+			return fmt.Errorf("value %f is out of range, greater or equal than maximum %f", value, min)
+		}
+	}
+	return nil
+}
+
+func validateRangeMaxFieldValue(max float64, value float64, inclusive bool) error {
+	if inclusive {
+		if max < value {
```
Comment (trapacska) [inline stepconf/stepconf.go]: same as above

File: stepconf/stepconf.go (hunk 8 above)
Comment (richard-bogdan-bitrise) [reply stepconf/stepconf.go]: Updated

File: stepconf/stepconf.go (hunk 9 above)
Comment (richard-bogdan-bitrise) [reply stepconf/stepconf.go]: Updated

File: stepconf/stepconf.go (hunk 10 above)
Comment (richard-bogdan-bitrise) [reply stepconf/stepconf.go]: Merged.

File: stepconf/stepconf.go (hunk 11 above)
Comment (richard-bogdan-bitrise) [reply stepconf/stepconf.go]: Merged.

**Pull request merged**